### DISCUSSION TITLE
Contributing

### DISCUSSION THREAD

**Discussion by gracekishino** [mdwiltfong/secret-santa#10] [General]
Hello! I'd love to contribute. I haven't used Node.js, Vite (React), TypeScript, Prisma or Express before, but have experience with PostgreSQL and Javascript. So something really simple to start with would be great!

**Comment by ivavay** [comment]
Hi! I'd love to contribute as well. I've used Vite (React) but not the other stacks. Would love to get exposed to the other tools as well but maybe I can get started with React first.

**Comment by mdwiltfong** [comment]
Oh my gosh, thank you both for reaching out! I'm excited :D 

@ivavay ,

Since you have some experience with React and Vite, do you mind making a PR for some mockups? A good place to start would be visualizing the gift giving session. You don't have the data yet because the backend is not where it needs to be to pull in users, but depending on the progress you can maybe mock that data too, and then by the time the backend is ready we can conform to your needs. You can use Postman or maybe even something like json-server: https://www.npmjs.com/package/json-server 

Here are the types: https://www.npmjs.com/package/@types/json-server

@gracekishino ,

Would you be able to look into setting up Prisma for authentication? This might be a challenge, but I think this article will really help out: https://www.prisma.io/blog/backend-prisma-typescript-orm-with-postgresql-auth-mngp1ps7kip4 A

**Reply by ivavay** [reply]
Okay! I'll do the mockups first and then work in React.

**Reply by gracekishino** [reply]
I'd love to have a go at setting up Prisma for authentication! 
I'll read through the article and let you know if it seems too challenging or not :)

**Comment by mdwiltfong** [comment]
I'll move the mockups conversation to this [discussion](https://github.com/mdwiltfong/secret-santa/discussions/12)

**Comment by freddieb77** [comment]
Hello! I'd love to contribute. I haven't used, Vite (React), TypeScript, Prisma or PostgreSQL.  I have the experience of working on the basics of  JavaScript, and React in the Codecademy curriculum and have started on the Basics in Redux. So something simple to start with would be great! I look forward to the challenge of working on and learning something new.

**Reply by mdwiltfong** [reply]
Hey @freddieb77 ! Thank you so much! 

We'll probably need to use Redux in the frontend. Do you want to make a PR that will help set that up? Here are some notes: 
- You can install the relevant packages in `/client`
- The frontend hasn't made enough progress yet on how we are going to use Redux. But would you be able to write some simple smoke tests? As in, tests that prove the installation of Redux is working? 
- It has to support TypeScript. This just means that you have to make sure you install the types. This [article](https://redux.js.org/usage/usage-with-typescript#standard-redux-toolkit-project-setup-with-typescript) will help.

**Reply by freddieb77** [reply]
@mdwiltfong,
Ok sounds great! I would like to clarify some things to make sure I am on the right track here.

For the smoke test, my plan was to write a simple test module that would that would increment/decrement a state by 1. Then write 3 tests that would check the initialState and then a dispatch test to increment and one to decrement. Funny enough this is what I was working on in my lesson today in Codecademy. 

If that is what you are looking for in the test then my next question is where would you want me to put the module. In the /client or make a Redux dir (/client/Redux) and place it there? Should the tests I write go into the /client/test/smoke.test.ts or do you want those test in a seperate file?

**Reply by mdwiltfong** [reply]
@freddieb77 ,

That sounds great. Will this test work even as we begin developing the state of the app? Just want to make sure the smoke test works regardless of how the app scales.

As for the structure, I noticed that Redux has a [vite+ts template](https://github.com/reduxjs/redux-templates/tree/master/packages/vite-template-redux). Here are some [tests](https://github.com/reduxjs/redux-templates/blob/master/packages/vite-template-redux/src/features/counter/counterSlice.spec.ts) that I found in that template. 

The workflow currently looks frontend tests that are under the `/tests` folder. So you should definitely put it there. Since it's a smoke test you can add it to `smoke.test.ts` 

As for the actual placement of the hooks, redux, etc i'm not so sure. I need to study redux some more. Maybe for now just make a `/redux` directory under `/client`. 

Let me know if that makes sense.

**Reply by mdwiltfong** [reply]
I just made an [issue](https://github.com/mdwiltfong/secret-santa/issues/20) of this. Just add a comment and I can assign it to you!

**Comment by MelindaH26** [comment]
Hello, I'm hoping I might have enough skills to help out with something small. I know JS, CSS, HTML and have done some work with Express routers, PostgreSQL, Node.js. I don't know TypeScript and have only just started learning React.

Are there any Issues that you could recommend for me to work on?

**Reply by mdwiltfong** [reply]
Hey @MelindaH26 ! 

Thank you so much for considering to contribute! 

I think you'd be perfect in setting up an [axios](https://axios-http.com/docs/example) instance for the front end. Basically React is going to be making a series of HTTP requests to the server. As the app grows, we don't want each component to be making it's own independent request. It would be easier if each component used a class, or function to make these requests. That way, if there is an error, it can be isolated to that single piece of code. 

Here is an [example](https://github.com/mdwiltfong/jamming-nx/blob/031f2d2caa75b0e74d263db0b51533e8c495794f/jamming-app/apps/jamming-front-end/src/app/helper_functions/APIHandler.tsx#L47) of a class I've made in the past. You can see it being used [here](https://github.com/mdwiltfong/jamming-nx/blob/031f2d2caa75b0e74d263db0b51533e8c495794f/jamming-app/apps/jamming-front-end/src/app/components/customHooks/useFetchUser.tsx#L14). 

We would want to do this a little differently though. Mainly we'd want to take advantage of creating an [axios instance](https://axios-http.com/docs/instance). 

Is this something you think you'd be willing to help out with? 

Michael

**Reply by mdwiltfong** [reply]
Also, I can definitely help you with TypeScript. That will be a smaller part of this contribution.

**Reply by MelindaH26** [reply]
Thanks - I will do some more research to try and get my head around exactly what an Axios is and give it a go. 

Could I start out with it in Javascript and then refactor it to TypeScript? Or would be better to do it as TypeScript from the start?

**Reply by mdwiltfong** [reply]
Sure! Basically `axios` is a library to help make writing HTTP requests easier. It's an upgrade to `fetch`

You can definitely write something up in JS and maybe next week we can put it all together. I'll make an issue and post it here. Just comment on it when you have a chance!

**Reply by mdwiltfong** [reply]
Here's the [issue](https://github.com/mdwiltfong/secret-santa/issues/48) @MelindaH26 ! If you don't mind leaving a comment there so I can assign it to you.
